logstash-core/lib/logstash/agent.rb

```diff
@@ -204,6 +250,72 @@ def close_pipelines
   end
 
   private
+  def transition_to_stopped
+    @running.make_false
+  end
+
+  def transition_to_running
+    @running.make_true
+  end
+
+  # We depends on a series of task derived from the internal state and what
+  # need to be run, theses actions are applied to the current pipelines to converge to
+  # the desired state.
+  #
+  # The current actions are simple and favor composition, allowing us to experiment with different
+  # way to making them and also test them in isolation with the current running agent.
+  #
+  # Currently only action related to pipeline exist, but nothing prevent us to use the same logic
+  # for other tasks.
+  def converge_state(pipeline_actions)
+    logger.debug("Converging pipelines")
+
+    converge_result = LogStash::ConvergeResult.new(pipeline_actions.size)
+
+    logger.debug("Needed actions to converge", :actions_count => pipeline_actions.size) unless pipeline_actions.empty?
+
+    pipeline_actions.each do |action|
+      # We execute every task we need to converge the current state of pipelines
+      # for every task we will record the action result, that will help us
+      # the results of all the task will determine if the converge was successful or not
+      #
+      # The ConvergeResult#add, will accept the following values
+      #  - boolean
+      #  - FailedAction
+      #  - SuccessfulAction
+      #  - Exception
+      #
+      # This give us a bit more extensibility with the current startup/validation model
+      # that we currently have.
+      begin
+        logger.debug("Executing action", :action => action)
+        action_result = action.execute(@pipelines)
+        converge_result.add(action, action_result)
```

Comment:
I could probably add a logger statement here when the `action fail`. I have mixed feeling if we should go with `debug` or `error`,